source/pele/pspin_spherical.h

```diff
@@ -0,0 +1,212 @@
+#ifndef _P_SPIN_SPHERICAL_H_
+#define _P_SPIN_SPHERICAL_H_
+
+#include "array.h"
+#include "base_potential.h"
+
+#include <iostream>
+#include <cassert>
+#include <cmath>
+#include "pele/meta_pow.h"
+#include "pele/combination.h"
+
+namespace pele {
+
+template <size_t p>
+class MeanFieldPSpinSpherical : public BasePotential {
+protected:
+    pele::Array<double> m_interactions, m_spins;
+    pele::Array<size_t> m_indexes;
+    size_t m_N;
+    double m_tol;
+    static const size_t m_p = p;
+    pele::combination_generator<size_t*> m_combination_generator;
+    pele::combination_generator<size_t*> m_combination_generator_grad;
+    pele::combination_generator<size_t*> m_combination_generator_hess;
+    size_t m_get_index(size_t* comb){
+        size_t idx = comb[0];
+        size_t d = m_N;
+        for (size_t i = 1; i < m_p; ++i){
+            idx += comb[i] * d;
+            d *= m_N;
+        }
+        return idx;
+    }
+
+    //this function normalizes the spin vector and orthogonalises the gradient to it
+    inline void m_orthonormalize(Array<double>& spins_, Array<double>& grad)
+    {
+        if (spins_.size() != grad.size()) {
+            throw std::invalid_argument("grad.size() be the same as spin_.size()");
+        }
+
+        bool success = true;
+        Array<double> spins = spins_.copy();
+
+        grad /= norm(grad);
```

Comment:
why are you normalizing the gradient?  this doesn't seem correct.  You want to subtract the component parallel to the spin vector.  Can't this whole function be replaced by the below line?

    grad -= dot(grad, spins) / norm(spins)